### PR TITLE
libzmq: fix build with Apple Clang 15.0.0

### PR DESCRIPTION
**Pull request by adamjstewart**
Apple Clang 15.0.0 also enables C++16 by default, causing many of the same issues GCC 13 did. Instead of iterating over all possible compilers, we should always apply all patches. These patches were merged into master because the developers trusted them not to break other compilers.

### CODE REVIEW

File: var/spack/repos/builtin/packages/libzmq/package.py [hunk 1]
```diff
@@ -70,23 +70,23 @@ class Libzmq(AutotoolsPackage):
 
     # Fix aggressive compiler warning false positive
     patch(
-        "https://github.com/zeromq/libzmq/commit/92b2c38a2c51a1942a380c7ee08147f7b1ca6845.patch?full_index=1",
+        "https://github.com/zeromq/libzmq/pull/4213.patch?full_index=1",
```
Comment (dennisklein) [inline var/spack/repos/builtin/packages/libzmq/package.py]: For my curiosity: What is the rationale behind this change?

**Review by dennisklein — APPROVED** [review]

File: var/spack/repos/builtin/packages/libzmq/package.py (hunk 1 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/libzmq/package.py]: Just a shorter URL. The patch itself is the same and we checksum the download so we don't need to specify the commit hash.

Comment (haampie) [reply var/spack/repos/builtin/packages/libzmq/package.py]: I would stick to the commit, idk you if you can push to a branch of a merged PR, but in principle a PR is a moving target

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/libzmq/package.py]: An open PR is a moving target (this has bit me in the past) but I don't think a merged PR is. I don't have strong opinions either way. If we decide Spack should have strong opinions we should add an audit check for this like we did for `?full_index=1` and other such things.

**Comment by adamjstewart** [comment]
@spackbot run pipeline

File: var/spack/repos/builtin/packages/libzmq/package.py (hunk 1 above)
Comment (dennisklein) [reply var/spack/repos/builtin/packages/libzmq/package.py]: I personally have no strong preference whether to use `pull/*.patch?full_index=1` or `commit/*.patch?full_index=1` urls. Once the project comes up with a policy here, I am happy to align.

Thx for the clarification. I just wished you put this rationale in the commit msg, or better made a separate commit. (But then this project **squash-merges** PRs anyways ... :crying_cat_face: :stuck_out_tongue_winking_eye:)

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/libzmq/package.py]: I'll revert the URL change and let y'all decide in a follow-up PR. I just want to be able to build libzmq 😄

**Review by dennisklein — APPROVED** [review]

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
> C++16

^^

**Review by haampie — APPROVED** [review]
